### ISSUE TITLE
listening port & upstream port error for nghttpx

### ISSUE THREAD

**Issue by fcicq** [nghttp2/nghttp2#898]
I guess #857 is an related issue.
I'm also confused that why I cant get the log from stderr. but thankfully I got the log in /var/log/syslog.

command line:
/usr/local/bin/nghttpx -s -b "127.0.0.1,8444;;proto=h2" -f "127.0.0.1,8118;no-tls"  --no-via
also tried --backend-address-family=IPv4 .
Note: listening port 127.0.0.1:3000 and upstream 127.0.0.1:80 are unexpected, this is the main problem.

Apr 19 18:26:51 meegoubuntu nghttpx[7042]: [NOTICE] Listening on 127.0.0.1:3000
Apr 19 18:26:51 meegoubuntu nghttpx[7042]: [NOTICE] Listening on 127.0.0.1:8118
Apr 19 18:26:51 meegoubuntu nghttpx[7042]: [NOTICE] Worker process [7043] spawned
Apr 19 18:26:51 meegoubuntu nghttpx[7043]: [NOTICE] [LISTEN:0x7ffe2ae8d080] Created worker thread #0
Apr 19 18:27:00 meegoubuntu nghttpx[7043]: [WARN] [DCONN:0x7f114c00d230] Backend connect failed; addr=127.0.0.1:80: errno=111 (shrpx_http_downstream_connection.cc:1415)
Apr 19 18:27:00 meegoubuntu nghttpx[7043]: [WARN] Could not connect 1 times in a row; sleep for 1.46433 seconds (shrpx_connect_blocker.cc:94)
Apr 19 18:27:21 meegoubuntu nghttpx[7043]: [WARN] [DCONN:0x7f114c012f20] Backend connect failed; addr=127.0.0.1:80: errno=111 (shrpx_http_downstream_connection.cc:1415)

**Comment by fcicq** [comment]
resolved. 
I found strace writes
open("/etc/nghttpx/nghttpx.conf", O_RDONLY) = 4
and that config is the ubuntu provided one. so --conf=/dev/null should work

please leave a notice that a config file is read from somewhere (unless conf is /dev/null)... so is it ok to leave the issue temporarily open?

**Comment by fcicq** [comment]
alternately, if config file is present, warns if the user uses -b / -f

**Comment by tatsuhiro-t** [comment]
The nghttpx manual says that --conf defaults to /etc/nghttpx/conf.
https://nghttp2.org/documentation/nghttpx.1.html#cmdoption-nghttpx--conf

**Comment by fcicq** [comment]
yes, I found that AFTER the issue resolved.several issue here:1 hidden --conf default_file (unexpected config load if unspecified, needs warning if config is loaded)2 hybrid config interfaces(warn if -b & -f after config file loaded, or even disable opt parsing if --conf is enabled)3 maybe introduce --longhelpleave only short introduction for backend & frontend & mode & conf, move all the current help messages to longhelp.---- On Wed, 19 Apr 2017 20:20:14 +0800  Tatsuhiro Tsujikawa<notifications@github.com> wrote ----The nghttpx manual says that --conf defaults to /etc/nghttpx/conf. https://nghttp2.org/documentation/nghttpx.1.html#cmdoption-nghttpx--conf  —You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by tatsuhiro-t** [comment]
It seems to me that the root cause of this issue is that user might not know the default configuration file is loaded automatically although it is clearly documented.  Other than that, the initial loading sequence is well defined for nghttpx.
I'm personally reluctant to add these rather complex warnings.
What about adding NOTICE level log to notify user that nghttpx is loading a configuration file?

**Comment by fcicq** [comment]
this is the config file included in ubuntu/debian default nghttp2-proxy installation.

Notice the "errorlog-syslog=yes". hopefully no one else would use -b / -f, otherwise the same strange issue would happen and get nothing output on stderr.

# Example 1
# Proxying an HTTP server on localhost:80 to localhost:3000 (no TLS)
frontend=127.0.0.1,3000;no-tls
backend=127.0.0.1,80
errorlog-syslog=yes
workers=1

**Comment by tatsuhiro-t** [comment]
My suggestion to log NOTICE level log is done before reading configuration file.  So it is written to stderr.

If you have any problem with the file which comes with the debian/ubuntu package, please report to the maintainer of the package.

**Comment by fcicq** [comment]
Do as your wish? I don't really mind as I wont be trapped by this kind of problem again.

another suggestion would be changing the help string.
--conf=<PATH>
Load configuration from <PATH>

to something like:
--conf=<PATH>
Load configuration from <PATH>, nghttpx will ALWAYS read the config file, if that is not preferred, use /dev/null

---

Writing to syslog is important for a daemon (started by start-stop-daemon?), so I dont think that default would be a problem, so no need to file a bug there.

**Comment by tatsuhiro-t** [comment]
Thank you.
Now it is a bit easier now: https://github.com/nghttp2/nghttp2/commit/c3f5f5ca361a94093c82f2bc4e7b87075c8e55ba and https://github.com/nghttp2/nghttp2/commit/911d12f7c4de2c67c16f969b176cca29dbbf7cd1
